cli/module_registry.go

```diff
@@ -176,46 +174,152 @@ func UpdateModuleCommand(c *cli.Context) error {
 	}
 
 	// If the namespace isn't set, modify the meta.json to set it (if available)
-	if moduleID.Namespace == "" {
+	if moduleID.Prefix == "" {
 		org, err := resolveOrg(client, orgIDArg, publicNamespaceArg)
 		if err != nil {
 			return err
 		}
 		if org.PublicNamespace != "" {
-			moduleID.Namespace = org.PublicNamespace
+			moduleID.Prefix = org.PublicNamespace
 			manifest.Name = moduleID.toString()
 			if err := writeManifest(manifestPath, manifest); err != nil {
 				return err
 			}
 			fmt.Fprintf(c.App.Writer, "\nUpdated %s to use the public namespace of %q which is %q\n",
-				manifestPath, org.Name, org.PublicNamespace)
+				defaultManifestFilename, org.Name, org.PublicNamespace)
 			fmt.Fprintf(c.App.Writer, "You no longer need to specify org_id or public_namespace\n")
 		}
 	}
 
 	return nil
 }
 
+// UploadModuleCommand runs the command to upload a new version of a module.
+func UploadModuleCommand(c *cli.Context) error {
+	manifestPathArg := c.String("module")
+	publicNamespaceArg := c.String("public_namespace")
+	orgIDArg := c.String("org_id")
+	nameArg := c.String("name")
+	versionArg := c.String("version")
+	platformArg := c.String("platform")
+	tarballPath := c.Args().First()
+	if c.Args().Len() > 1 {
+		return errors.New("Too many arguments passed to upload command. " +
+			"Make sure to specify flag and optional arguments before the required positional package argument")
+	}
+	if tarballPath == "" {
+		return errors.New("No package to upload -- please provide an archive containing your module. See the help for more information")
+	}
+
+	client, err := NewAppClient(c)
+	if err != nil {
+		return err
+	}
+
+	manifestPath := defaultManifestFilename
+	if manifestPathArg != "" {
+		manifestPath = manifestPathArg
+	}
+	var moduleID ModuleID
+	// if the manifest cant be found
+	if _, err := os.Stat(manifestPath); err != nil {
```

Comment:
can we refactor this to be a separate function as this main function is pretty long and a little bit hard to follow. Something like:

```
func getModuleIDFromManifest(manifestPath) {
// if file does not exist return that and then do all this stuff
// otherwise will return err or ID in which case just error check that it is same as nameArg

}

```

Reply:
This is another thing that APP-2230 will fix. Once I can pass in a few args and get a `ModuleID` that actually works (whether it is namespace or orgid), things will be greatly simplified.

If you feel strongly that it should be done now, happy to do it. However this logic is going to change anyways in the next few weeks

Reply:
Ah, i saw that but didn't realize that that was encompassing organizational changes to the code too. If this will be done soon / this PR is a big blocker you can postpone this to the next PR. I do think this is a lil bit messy, though, so should be addressed or at least made clearer sooner rather than later especially as its important code.

Reply:
>  I do think this is a lil bit messy

100% and the fact that my name is on the git blame is plenty encouragement to get it polished up. 

I just want to get this running so that other people can test uploading modules. Then we can do a real cleanup (ex: we should have product sign off on all of the error messages that I just randomly wrote)

Reply:
okay sounds good -- leave for now :)